### PR TITLE
pkg/kvstore: add gRPC keep alives for etcd connectivity

### LINKED ISSUE #12945: Pod can be stuck in ContainerCreating state for ~15m with identity-allocation-mode=kvstore
## Bug report

A Pod remains stuck in ContainerCreating state for up to 15m with identity-allocation-mode=kvstore. After ~15m, the pod transitions to Running state, and everything proceeds normally. This delay (presumably in identity allocation) occurs when at least one etcd node is down. If the etcd cluster has quorum (e.g. 2/3), then this delay should not happen. 

**General Information**

- 1.7.7 & 1.8.2
- K8s with etcd backends
- Sysdumps:
  - [cilium-sysdump-20200820-185312-1.8.2.zip](https://github.com/cilium/cilium/files/5106796/cilium-sysdump-20200820-185312-1.8.2.zip)
  - [cilium-sysdump-20200820-181830-1.7.7-heartbeats-after-pod-running.zip](https://github.com/cilium/cilium/files/5106798/cilium-sysdump-20200820-181830-1.7.7-heartbeats-after-pod-running.zip)
   - [cilium-sysdump-20200820-181535-1.7.7-heartbeats-during-containercreating.zip](https://github.com/cilium/cilium/files/5106799/cilium-sysdump-20200820-181535-1.7.7-heartbeats-during-containercreating.zip)

**How to reproduce the issue**

1. Deploy K8s cluster with etcd nodes
2. Deploy Cilium with the following config:
    -  `kvstore: etcd`
    - `identity-allocation-mode: kvstore`
3. Block one etcd node via https://gist.github.com/christarazi/11aadf01d353112eb10ed82373569155. Note this **must** be done before deploying the pods in the next step.
4. Deploy nginx (ensure that at least one replica lands on the Cilium instance that you've applied the above etcd iptables rules to):

```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx-deployment
  labels:
    app: nginx
    cool: stuff
spec:
  replicas: 3
  selector:
    matchLabels:
      app: nginx
  template:
    metadata:
      labels:
        app: nginx
    spec:
      containers:
      - name: nginx
        image: nginx:1.14.2
        ports:
        - containerPort: 80
```

### PR DESCRIPTION
**Pull request by aanm**
If the client does not receive a keep alive from the server, that
connection should be closed so the etcd client library does proper
round robin for the other available endpoints.

This might be a little bit aggressive in a larger environment if all
clients perform a keep alive requests to the etcd servers. Some
testing could be done to verify if there is a large overhead of
doing these keep alive requests.

Signed-off-by: André Martins <andre@cilium.io>

Fixes https://github.com/cilium/cilium/issues/12945

**TODO** Perhaps it would also be a good idea to do have a hidden flag to configure this timeouts? Or even disable them?

```release-note
Improved reliability of etcd connectivity by adding gRPC keep alives
```

### CODE REVIEW

**Review by joestringer — APPROVED** [review]

File: pkg/kvstore/etcd.go [hunk 1]
```diff
@@ -645,6 +680,11 @@ func connectEtcdClient(ctx context.Context, config *client.Config, cfgPath strin
 	// block until DialTimeout is reached or a connection to the server
 	// is made.
 	config.DialTimeout = 0
+	// Ping the server to verify if the server connection is still valid
+	config.DialKeepAliveTime = clientOptions.KeepAliveHeartbeat
+	// Timeout if the server does not reply within 15 seconds and close the
+	// connection. Ideally it should be lower than staleLockTimeout
+	config.DialKeepAliveTimeout = clientOptions.KeepAliveTimeout
```
Comment (joestringer) [inline pkg/kvstore/etcd.go:687]: Relating to the large-scale cluster concern, we could consider adjusting this based on the cluster size like we do for some other time-based checks in Cilium?

**Review by christarazi — APPROVED** [review]

**Review by joestringer — COMMENTED** [review]
One more question came up during discussion with @christarazi: Should we only enable this for etcd client connections to the local etcd instance? Ie in clustermesh, avoid configuring this option?

File: pkg/kvstore/etcd.go (hunk 1 above)
Comment (joestringer) [reply pkg/kvstore/etcd.go:687]: Something like `ClusterSizeDependantInterval`.

**Comment by tgraf** [comment]
> One more question came up during discussion with @christarazi: Should we only enable this for etcd client connections to the local etcd instance? Ie in clustermesh, avoid configuring this option?

I would treat them in the same way in general, the failover/stale connection problematic is exactly the same.

**Pull request merged**